pkg/agent/openflow/cookie/allocator.go

```diff
@@ -0,0 +1,129 @@
+// Copyright 2019 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cookie
+
+import (
+	"fmt"
+
+	"k8s.io/klog"
+)
+
+const (
+	BitwidthRound    = 16
+	BitwidthCategory = 8
+	BitwidthReserved = 64 - BitwidthCategory - BitwidthRound
+
+	RoundMask    = uint64(0xffff000000000000)
+	CategoryMask = uint64(0x0000ff0000000000)
+)
+
+// Category represents the flow entry category.
+type Category uint64
+
+const (
+	Default Category = iota
+	Gateway
+	Node
+	Pod
+	Service
+	Policy
+)
+
+func (c Category) String() string {
+	switch c {
+	case Default:
+		return "Default"
+	case Gateway:
+		return "Gateway"
+	case Node:
+		return "Node"
+	case Pod:
+		return "Pod"
+	case Service:
+		return "Service"
+	case Policy:
+		return "Policy"
+	default:
+		return "Invalid"
+	}
+}
+
+// ID defines segments a cookie ID contains. An ID is composed like:
+//  |------------------------- ID --------------------------|
+//  |- round 16bits -|- category 8bits -|- reserved 40bits -|
+// The round segment represents the round id.
+// The category segment represents the category of flow this ID belongs.
+type ID uint64
+
+func newID(round uint64, cat Category) ID {
+	r := uint64(0)
+	r |= round << (64 - BitwidthRound)
+	r |= (uint64(cat) << BitwidthReserved) & CategoryMask
+	return ID(r)
+}
+
+// Raw returns the unit64 type value of the ID.
+func (i ID) Raw() uint64 {
+	return uint64(i)
+}
+
+// Round returns the round number of the ID.
+func (i ID) Round() uint64 {
+	return i.Raw() >> (64 - BitwidthRound)
+}
+
+// Category returns the category of the ID.
+func (i ID) Category() Category {
+	return Category((i.Raw() & CategoryMask) >> BitwidthReserved)
+}
+
+// String returns the string representation of the ID.
+func (i ID) String() string {
+	return fmt.Sprintf("<round:%d,category:%s>", i.Round(), i.Category().String())
+}
+
+// Allocator defines operations of a cookie ID allocator.
+type Allocator interface {
+	// Request cookie IDs of flow categories.
+	Request(cat Category) ID
+}
+
+type allocator struct {
+	round        uint64
+	preallocated map[Category]ID // TODO: since IDs of each category in the same round will not be changed, pre-allocating them to gain some performance
```

Comment:
I don't understand this TODO. You are already populating the map in `NewAllocator`, so it seems there is nothing left to do. The conversation with @tnqn was more about changing the allocate / request terminology.

Also I wouldn't mention "gain some performance" at all here. Personally I'm not sure that doing a map lookup is faster than doing some bit arithmetic. And even if there is a difference, it's probably irrelevant.

I think the current implementation is fine BTW (with a map lookup instead of some variables exposed directly to the client).  It's better IMO to use a method so we can change the implementation later if desired.

Reply:
Thanks, I will remove the TODO.
BTW, just tested performance between map and the original way. For 1,000,000,000 requests in sequential,
BitArithmetic way: 2.229329272s
Map way: 5.753172249s
I will revert the map implementation.